File: share/example/python/example_timer.py
#! /usr/bin/env python
# coding=utf-8

"""
Test timer functions: start, stop, reset, delete

Steps:
Step 1: Connect to RPC service, robot login
Step 2: Example 1
    1. Timer timer1 starts timing, after 7.5 s, timer timer1 stops, get duration about 7.5s
    2. Repeat the above step, get duration about 15.0s
    3. Reset timer timer1, get duration about 0.0s, delete timer timer1
Step 3: Example 2
    1. Timer timer1 starts timing, after 7.5 s, get duration about 7.5s, reset timer timer1
    2. Timer timer1 starts timing, after 7.5 s, get duration about 7.5s, reset and delete timer timer1
"""

import time
import pyaubo_sdk

robot_ip = "127.0.0.1"  # Server IP address
robot_port = 30004  # Port number
M_PI = 3.14159265358979323846
robot_rpc_client = pyaubo_sdk.RpcClient()

if __name__ == '__main__':
    robot_rpc_client.connect(robot_ip, robot_port)  # API call: connect to RPC service
    if robot_rpc_client.hasConnected():
        print("Robot rcp_client connected successfully!")
        robot_rpc_client.login("aubo", "123456")  # API call: robot login
        if robot_rpc_client.hasLogined():
            print("Robot rcp_client logined successfully!")
            robot_name = robot_rpc_client.getRobotNames()[0]  # API call: get robot name
            impl = robot_rpc_client.getRobotInterface(robot_name)

            print("Example 1---------------")
            # Timer timer1 starts timing, after 7.5 s, timer timer1 stops, get duration about 7.5s
            robot_rpc_client.getRuntimeMachine().timerStart("timer1")
            time.sleep(7.5)
            robot_rpc_client.getRuntimeMachine().timerStop("timer1")
            timer1 = robot_rpc_client.getRuntimeMachine().getTimer("timer1")
            print("Duration = ", timer1)

            # Timer timer1 starts timing, after 7.5 s, timer timer1 stops, get duration about 15.0s
            robot_rpc_client.getRuntimeMachine().timerStart("timer1")
            time.sleep(7.5)
            robot_rpc_client.getRuntimeMachine().timerStop("timer1")
            timer1 = robot_rpc_client.getRuntimeMachine().getTimer("timer1")
            print("Duration = ", timer1)

            # Reset timer timer1, get duration about 0.0s, delete timer timer1
            robot_rpc_client.getRuntimeMachine().timerReset("timer1")
            timer1 = robot_rpc_client.getRuntimeMachine().getTimer("timer1")
            print("Duration = ", timer1)
            robot_rpc_client.getRuntimeMachine().timerDelete("timer1")

            print("Example 2---------------")
            # Timer timer1 starts timing, after 7.5 s, get duration about 7.5s, reset timer timer1
            robot_rpc_client.getRuntimeMachine().timerStart("timer1")
            time.sleep(7.5)
            timer1 = robot_rpc_client.getRuntimeMachine().getTimer("timer1")
            print("Duration = ", timer1)
            robot_rpc_client.getRuntimeMachine().timerReset("timer1")

            # Timer timer1 starts timing, after 7.5 s, get duration about 7.5s, reset and delete timer timer1
            robot_rpc_client.getRuntimeMachine().timerStart("timer1")
            time.sleep(7.5)
            timer1 = robot_rpc_client.getRuntimeMachine().getTimer("timer1")
            print("Duration = ", timer1)
            robot_rpc_client.getRuntimeMachine().timerReset("timer1")
            robot_rpc_client.getRuntimeMachine().timerDelete("timer1")

File: share/example/c++/example_timer.cpp
#include "aubo_sdk/rpc.h"
#include <math.h>
#include <thread>

using namespace arcs::common_interface;
using namespace arcs::aubo_sdk;

#define LOCAL_IP "127.0.0.1"

void example1(RpcClientPtr rpc_cli) {
  // API call: Timer timer1 starts timing
  rpc_cli->getRuntimeMachine()->timerStart("timer1");
  std::this_thread::sleep_for(std::chrono::milliseconds(7500));
  // API call: Timer timer1 stops timing
  rpc_cli->getRuntimeMachine()->timerStop("timer1");
  // API call: Get the duration timed by timer timer1
  std::cout << rpc_cli->getRuntimeMachine()->getTimer("timer1") << std::endl;

  // API call: Timer timer1 starts timing
  rpc_cli->getRuntimeMachine()->timerStart("timer1");
  std::this_thread::sleep_for(std::chrono::milliseconds(7500));
  // API call: Timer timer1 stops timing
  rpc_cli->getRuntimeMachine()->timerStop("timer1");
  // API call: Get the duration timed by timer timer1
  std::cout << rpc_cli->getRuntimeMachine()->getTimer("timer1") << std::endl;

  // API call: Reset timer timer1
  rpc_cli->getRuntimeMachine()->timerReset("timer1");
  std::this_thread::sleep_for(std::chrono::milliseconds(50));
  // API call: Get the duration timed by timer timer1
  std::cout << rpc_cli->getRuntimeMachine()->getTimer("timer1") << std::endl;
  // API call: Delete timer timer1
  rpc_cli->getRuntimeMachine()->timerDelete("timer1");
}

void example2(RpcClientPtr rpc_cli) {
  // API call: Timer timer1 starts timing
  rpc_cli->getRuntimeMachine()->timerStart("timer1");
  std::this_thread::sleep_for(std::chrono::milliseconds(7500));
  // API call: Get the duration timed by timer timer1
  std::cout << rpc_cli->getRuntimeMachine()->getTimer("timer1") << std::endl;
  // API call: Reset timer timer1
  rpc_cli->getRuntimeMachine()->timerReset("timer1");

  // API call: Timer timer1 starts timing
  rpc_cli->getRuntimeMachine()->timerStart("timer1");
  std::this_thread::sleep_for(std::chrono::milliseconds(7500));
  // API call: Get the duration timed by timer timer1
  std::cout << rpc_cli->getRuntimeMachine()->getTimer("timer1") << std::endl;
  // API call: Reset timer timer1
  rpc_cli->getRuntimeMachine()->timerReset("timer1");
  // API call: Delete timer timer1
  rpc_cli->getRuntimeMachine()->timerDelete("timer1");
}

int main(int argc, char **argv) {
  auto rpc_cli = std::make_shared<RpcClient>();
  // API call: Set RPC timeout
  rpc_cli->setRequestTimeout(1000);
  // API call: Connect to RPC service
  rpc_cli->connect(LOCAL_IP, 30004);
  // API call: Login
  rpc_cli->login("aubo", "123456");

  example1(rpc_cli);
  example2(rpc_cli);

  // API call: Logout
  rpc_cli->logout();
  // API call: Disconnect
  rpc_cli->disconnect();

  return 0;
}
